### PR TITLE
refactor: report all diffs for SQLInstance

### PR DESCRIPTION
**Pull request by acpana**
### BRIEF Change description

Realized we are not actually reporting ALL the diffs for a SQLInstance.

Fixes https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/5419

#### WHY do we need this change?

It will improve the UX of preview, for one, and allow operators to see what the differences are between SQLInstance reconciles.

#### Does this PR add something which needs to be 'release noted'?

```release-note
Log all field diffs for `SQLInstance`.
```

- [ ] Reviewer reviewed release note.

#### Additional documentation e.g., references, usage docs, etc.:

```docs

```

#### Intended Milestone

Please indicate the intended milestone. 
- [ ] Reviewer tagged PR with the actual milestone.

Tagged with 1.140

### Tests you have done

- [ ] Run `make ready-pr` to ensure this PR is ready for review.
- [ ] Perform necessary E2E testing for changed resources.

### CODE REVIEW

File: pkg/controller/direct/sql/sqlinstance_equality.go [hunk 1]
```diff
@@ -22,59 +22,52 @@ import (
...
-func InstancesMatch(desired *api.DatabaseInstance, actual *api.DatabaseInstance, diff *structuredreporting.Diff) bool {
+func DiffInstances(desired *api.DatabaseInstance, actual *api.DatabaseInstance) *structuredreporting.Diff {
+	diff := &structuredreporting.Diff{}
 	if desired == nil && actual == nil {
-		return true
+		return diff
 	}
 	if !PointersMatch(desired, actual) {
-		return false
+		return diff
```
Comment (cheftako) [inline pkg/controller/direct/sql/sqlinstance_equality.go]: PointersMatch is a terrible name for that method. At this point we know that exactly one of desired/actual are not nil. Returning an empty diff does not seem correct. Also suspicious this does not conform to the old behavior.

Comment (cheftako) [reply pkg/controller/direct/sql/sqlinstance_equality.go]: Would be good to add a test for each of desired and actual being nil.

File: pkg/controller/direct/sql/sqlinstance_equality.go [hunk 2]
```diff
@@ -22,59 +22,49 @@ import (
...
 	if !ReplicaConfigurationsMatch(desired.ReplicaConfiguration, actual.ReplicaConfiguration) {
 		diff.AddField(".replicaConfiguration", actual.ReplicaConfiguration, desired.ReplicaConfiguration)
-		return false
 	}
 	// Ignore ReplicationCluster. It is not supported in KRM API.
 	// Ignore RootPassword. It is not exported.
-	if !SettingsMatch(desired.Settings, actual.Settings, diff) {
-		return false
-	}
+	diff.AddDiff(DiffSettings(desired.Settings, actual.Settings))
```
Comment (anhdle-sso) [inline pkg/controller/direct/sql/sqlinstance_equality.go:61]: Why don't you pass the diff pointer to this function and keep adding to the same obj? You would not have to call AddDiff

Comment (acpana) [reply pkg/controller/direct/sql/sqlinstance_equality.go:61]: I think @cheftako wanted us to move away from side effects like that 😛

File: pkg/controller/direct/sql/sqlinstance_equality.go (hunk 1 above)
Comment (acpana) [reply pkg/controller/direct/sql/sqlinstance_equality.go]: I've reworked this to remove the first pointers match!

File: pkg/controller/direct/sql/sqlinstance_equality.go (hunk 2 above)
Comment (cheftako) [reply pkg/controller/direct/sql/sqlinstance_equality.go:61]: +1. If a function name makes it clear that it mutates its inputs then we can discuss it. However the function DiffSettings suggests it does not mutate its inputs. AddDiff mya not modify its inputs but its at least clear its accumulating a result.
